### ISSUE TITLE
NaN's in saved model only when training with tensorflow-gpu

### ISSUE THREAD

**Issue by clkruse** [tensorflow/tensorflow#38698]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  I believe Debian 9 "Stretch" on google cloud and whatever distro Google colab uses.
- TensorFlow installed from (source or binary): Installed as binary
- TensorFlow version (use command below): v2.1.0-rc2-17-ge5bf8de 2.1.0 and v2.2.0-rc2-77-gaad398b5e9 2.2.0-rc3
- Python version: 3.7
- CUDA/cuDNN version: Uncertain
- GPU model and memory: Tesla V100 on google cloud, uncertain on colab

**Describe the current behavior**
I have been building off of the [tensorflow pix2pix example](https://www.tensorflow.org/tutorials/generative/pix2pix).

Steps:
1. Train the network and save checkpoints along the way. Training proceeds normally with reasonable losses, and intermediate outputs are look correct. 

2. Reload the network from a checkpoint and export the generator as an .h5: `generator.save('generator.h5', save_format="h5")`.

3. Reload the saved h5 network using `tf.keras.models.load_model('generator.h5')`

4. If the network was trained with tensorflow-gpu, a block of weights on one layer (only part of sequential_9) will be NaN. If the network was trained with tensorflow cpu, the weights and outputs are appropriate.

**Describe the expected behavior**

I expect the saved and loaded network to behave in the same way as the network as it is training, and the weights to not contain any nan values.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

Apologies for including the full pix2pix script. I'll attempt to slim down the code later. Here is a colab notebook that contains the pix2pix script, and the weights from the `sequential_9` layer trained on tf-gpu and tf-cpu. To reproduce, change the runtime type between GPU and CPU.
https://colab.research.google.com/drive/1im9dVf63vuldIGsnQ48sHBKHHpDVuHvG

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by clkruse** [comment]
It seems like this may be the same issue present in #35307. Unfortunately, I don't see a way to pass `clear_devices=True` in TF2

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.2.0rc3](https://colab.research.google.com/gist/amahendrakar/3ea664961193083b74e8866646e7c386/38698-gpu.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/3560262fb4639f439943e9a183af3f0f/38698-tf-nightly.ipynb). Please find the attached gist. Thanks!

**Comment by clkruse** [comment]
I'm making a little progress in troubleshooting. It looks like the issue is coming from the BatchNormalization layer in the upsample block. If I comment out the BatchNorm layer in the upsampling function (line 129 in the [linked gist](https://colab.research.google.com/drive/1im9dVf63vuldIGsnQ48sHBKHHpDVuHvG)), there are no NaNs in the saved model.

**Comment by clkruse** [comment]
Isolating BatchNorm allowed me to find related issues. In this issue #29478, I noticed that the original code for generating predictions from the network uses `training=True` as an argument to the prediction. When the model is called, the network no longer produces NaN values with this argument.

However, the issue with the NaN weights in the saved model persists. This means that the saved model cannot be used in TensorFlow JS as TFJS does not have a `training=true` function that I am aware of.

**Comment by StevenBanama** [comment]
You can clean it by hand as below. It works for me.
models.load_weights("{path}")
model_refresh_without_nan(models)

def model_refresh_without_nan(models):
    import numpy as np
    valid_weights = []
    for l in models.get_weights():
        if np.isnan(l).any():
            valid_weights.append(np.nan_to_num(l))
            print("!!!!!", l)
        else:
            valid_weights.append(l)
    models.set_weights(valid_weights)

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tensorflow GPU 2.5, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/b8520107f45c3500954e2593ec047e61/38698.ipynb). Thanks!

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!
